python/pyarrow/io.pxi

```diff
@@ -1449,6 +1449,10 @@ cdef shared_ptr[CBuffer] as_c_buffer(object o) except *:
 
 
 cdef NativeFile get_native_file(object source, c_bool use_memory_map):
+    if source is None:
```

Comment:
I'm not sure this is useful. Why would you pass `None` in the first place?
Not crashing is nice, but I don't think we want to maintain a specific error message for this case.

Reply:
I suggested to add `object source not None` in the signature (that would automatically give a better error message), but so apparently that only works for `def` functions and not for `cdef` functions.

Reply:
Yes, I can confirm that `not None` is not allowed in `cdef`  functions:

```
pyarrow/io.pxi:1451:32: 'not None' only allowed in Python functions
```

I think that the main value probably comes from having a test to cover for a past `segfault`, the better error message is a nice to have but not strictly something that values the extra code. If we want to go deeper we might want to investigate where the segfault was originally coming from (as my feeling is that we have just hidden it behind other python changes) but it doesn't seem to value the extra time at the moment if it's already solved.

Reply:
@amol- another solution could be to add `not None` in places where `get_native_file` eventually gets called, to get an error higher up. Eg `ParquetReader.open` or `ORCReader.open` could have `object source not None`, since those are `def` methods.

Reply:
The same issue seems to involved btw also callers of `get_input_stream` which is a bit more widespread.
See for example
```
>>> json.read_json(None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pyarrow/_json.pyx", line 238, in pyarrow._json.read_json
    _get_reader(input_file, &stream)
  File "pyarrow/_json.pyx", line 193, in pyarrow._json._get_reader
    get_input_stream(input_file, use_memory_map, out)
  File "pyarrow/io.pxi", line 1494, in pyarrow.lib.get_input_stream
    input_stream = nf.get_input_stream()
AttributeError: 'NoneType' object has no attribute 'get_input_stream'
```

I wonder if a possible strategy might also be to provide an error signaling return value to `get_reader` and `get_input_stream` (as they don't return anything) and use it for the purpose of adding an `except -1` or something like that.

That seems would cover more paths that could all end up in this situation.

Reply:
Again, I'm not sure what the concern is. Is it common to pass `None` instead of a file object? It's not an optional parameter, so I don't know where the `None` could come from.

(also, getting this kind of `AttributeError` when passing the wrong type is quite common in Python)